Bottom: c7daca87c81a4683e64300a34d62f38021a142cd
Top:    cc28380b14236a729e8d554d8ff7b124aa53ba20
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-05-11 12:14:21 +0100

gtk/gl: Play with layer translucency


---

diff --git a/src/hid/gtk/gtkhid-gl.c b/src/hid/gtk/gtkhid-gl.c
index ccb4391..eb0a5e1 100644
--- a/src/hid/gtk/gtkhid-gl.c
+++ b/src/hid/gtk/gtkhid-gl.c
@@ -666,10 +666,11 @@ ghid_fill_pcb_polygon (hidGC gc, PolygonType *poly, const BoxType *clip_box)
 void
 ghid_thindraw_pcb_polygon (hidGC gc, PolygonType *poly, const BoxType *clip_box)
 {
+  double old_global_alpha_mult = global_alpha_mult;
   common_thindraw_pcb_polygon (gc, poly, clip_box);
-  ghid_global_alpha_mult (gc, 0.25);
+  ghid_global_alpha_mult (gc, old_global_alpha_mult * 0.25);
   ghid_fill_pcb_polygon (gc, poly, clip_box);
-  ghid_global_alpha_mult (gc, 1.0);
+  ghid_global_alpha_mult (gc, old_global_alpha_mult);
 }
 
 void
@@ -1698,6 +1699,7 @@ void
 ghid_draw_everything (BoxTypePtr drawn_area)
 {
   int i, ngroups;
+  int number_phys_on_top;
   int side;
   /* This is the list of layer groups we will draw.  */
   int do_group[MAX_LAYER];
@@ -1784,26 +1786,39 @@ ghid_draw_everything (BoxTypePtr drawn_area)
   }
 
   /* draw all layers in layerstack order */
+#define FADE_FACTOR 0.6
+  number_phys_on_top = max_phys_group - min_phys_group;
   for (i = ngroups - 1; i >= 0; i--) {
+    bool is_this_physical = drawn_groups[i] >= min_phys_group &&
+                            drawn_groups[i] <= max_phys_group;
+    bool is_next_physical = i > 0 &&
+                            drawn_groups[i - 1] >= min_phys_group &&
+                            drawn_groups[i - 1] <= max_phys_group;
+
+    double alpha_mult = global_view_2d ? pow (FADE_FACTOR, i) :
+      (is_this_physical ? pow (FADE_FACTOR, number_phys_on_top) : 1.);
+
+    if (is_this_physical)
+      number_phys_on_top --;
+
+    ghid_global_alpha_mult (Output.fgGC, alpha_mult);
     DrawLayerGroup (drawn_groups [i], drawn_area);
 
 #if 1
-    if (!global_view_2d && i > 0 &&
-        drawn_groups[i] >= min_phys_group &&
-        drawn_groups[i] <= max_phys_group &&
-        drawn_groups[i - 1] >= min_phys_group &&
-        drawn_groups[i - 1] <= max_phys_group) {
+    if (!global_view_2d && is_this_physical && is_next_physical) {
       cyl_info.from_layer = drawn_groups[i];
       cyl_info.to_layer = drawn_groups[i - 1];
       cyl_info.scale = gport->zoom;
       gui->set_color (Output.fgGC, "drill");
-      ghid_global_alpha_mult (Output.fgGC, 0.75);
+      ghid_global_alpha_mult (Output.fgGC, alpha_mult * 0.75);
       if (PCB->PinOn) r_search (PCB->Data->pin_tree, drawn_area, NULL, pin_hole_cyl_callback, &cyl_info);
       if (PCB->ViaOn) r_search (PCB->Data->via_tree, drawn_area, NULL, via_hole_cyl_callback, &cyl_info);
-      ghid_global_alpha_mult (Output.fgGC, 1.0);
     }
 #endif
   }
+#undef FADE_FACTOR
+
+  ghid_global_alpha_mult (Output.fgGC, 1.0);
 
   if (TEST_FLAG (CHECKPLANESFLAG, PCB))
     return;
